[macOS] Make it possible to test smart magnify in layout tests

smfr opened the pull request:
#### 835f5fc2d0ec6def0f44cfc3d343d7f45cc82588
<pre>
[macOS] Make it possible to test smart magnify in layout tests
<a href="https://bugs.webkit.org/show_bug.cgi?id=266449">https://bugs.webkit.org/show_bug.cgi?id=266449</a>
<a href="https://rdar.apple.com/119699103">rdar://119699103</a>

Reviewed by Tim Horton.

Make it possible to write a &quot;smart magnify&quot; test. There are several components to this:
1. We teach EventSender how to generate the event, which is an NSEvent of type NSEventTypeSmartMagnify
   with kIOHIDEventTypeZoomToggle.
2. We allow a test to enable WKWebView magnification via UIScriptController. We also fix WebKitTestRunner
   to reset the magnification state between tests.
3. We give WheelEventTestMonitor a defer reason so that a test can wait for the zoom animation to complete
   (yes, this isn&apos;t wheel events, but the infrastructure is exactly what we need).
   To support this, DrawingArea::commitTransientZoom() returns an async reply, and
   RemoteLayerTreeDrawingAreaProxyMac::commitTransientZoom() sets up the deferral reason across the
   CAAnimation, and the reply from the web process that the zoom is committed.

A new test is added in LayoutTests/view-gestures/smart-magnify. In future, we should move the swipe
tests into this view-gestures directory.

* LayoutTests/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/resources/ui-helper.js:
(window.UIHelper.async smartMagnifyAt):
(window.UIHelper.setWebViewAllowsMagnification):
* LayoutTests/view-gestures/resources/view-gesture-helpers.js: Added.
(visualViewportDescription):
* LayoutTests/view-gestures/smart-magnify/double-tap-zoom-expected.txt: Added.
* LayoutTests/view-gestures/smart-magnify/double-tap-zoom.html: Added.
* Source/WebCore/page/WheelEventTestMonitor.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/page/WheelEventTestMonitor.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::commitTransientZoom):
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::sendCommitTransientZoom):
* Source/WebKit/UIProcess/mac/TiledCoreAnimationDrawingAreaProxy.mm:
(WebKit::TiledCoreAnimationDrawingAreaProxy::commitTransientZoom):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/DrawingAreaCoordinatedGraphics.cpp:
(WebKit::DrawingAreaCoordinatedGraphics::commitTransientZoom):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/DrawingAreaCoordinatedGraphics.h:
* Source/WebKit/WebProcess/WebPage/DrawingArea.h:
(WebKit::DrawingArea::commitTransientZoom):
* Source/WebKit/WebProcess/WebPage/DrawingArea.messages.in:
* Source/WebKit/WebProcess/WebPage/mac/RemoteLayerTreeDrawingAreaMac.h:
* Source/WebKit/WebProcess/WebPage/mac/RemoteLayerTreeDrawingAreaMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaMac::commitTransientZoom):
* Source/WebKit/WebProcess/WebPage/mac/TiledCoreAnimationDrawingArea.h:
* Source/WebKit/WebProcess/WebPage/mac/TiledCoreAnimationDrawingArea.mm:
(WebKit::TiledCoreAnimationDrawingArea::commitTransientZoom):
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
(WTR::UIScriptController::setWebViewAllowsMagnification):
* Tools/WebKitTestRunner/EventSenderProxy.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/EventSendingController.idl:
* Tools/WebKitTestRunner/InjectedBundle/EventSendingController.cpp:
(WTR::EventSendingController::smartMagnify):
* Tools/WebKitTestRunner/InjectedBundle/EventSendingController.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::didReceiveSynchronousMessageFromInjectedBundle):
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::cocoaResetStateToConsistentValues):
* Tools/WebKitTestRunner/mac/EventSenderProxy.mm:
(-[EventSenderSyntheticEvent initSmartMagnifyEventAtLocation:globalLocation:time:eventNumber:window:]):
(WTR::EventSenderProxy::smartMagnify):
* Tools/WebKitTestRunner/mac/UIScriptControllerMac.h:
* Tools/WebKitTestRunner/mac/UIScriptControllerMac.mm:
(WTR::UIScriptControllerMac::setWebViewAllowsMagnification):

Canonical link: <a href="https://commits.webkit.org/272128@main">https://commits.webkit.org/272128@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5ab81796ff807c0f8354079f42a9de08799696d1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/30681 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/9362 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/32344 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/33179 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/27737 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/31391 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/11661 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/6619 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/27629 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/30993 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WK2-Tests-EWS "Waiting to run tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/27471 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/6733 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/27337 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/34517 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/27841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/33053 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/6916 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/5018 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/30879 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/8639 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/27114 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7257 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/7640 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/7469 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on LayoutTests/resources/ui-helper.js:
```diff
@@ -1255,6 +1255,22 @@ window.UIHelper = class UIHelper {
+    static async smartMagnifyAt(x, y)
+    {
+        if (!this.isWebKit2() || !this.isMac()) {
+            console.log('Smart magnify is only supported on macOS');
```
Err, not true as a feature. Maybe true for this function? Can this be worded more like "we haven't implemented it in the test runner?"

smfr replied:
On iOS I think you could trigger it with the existing touch event synthesis. I suppose this code could generate the appropriate touch events there.

aestes commented on LayoutTests/resources/ui-helper.js:
```diff
@@ -1255,6 +1255,22 @@ window.UIHelper = class UIHelper {
+
+        await UIHelper.startMonitoringWheelEvents();
+
+		// If smartMagnify is not working, ensure you've called setWebViewAllowsMagnification(true).
```
tabs

aestes commented on LayoutTests/view-gestures/smart-magnify/double-tap-zoom.html:
```diff
@@ -0,0 +1,44 @@
+	<div id="output"></div>
+	<div class="target"></div>
+</body>
+</html>
```
many tabs here. maybe we don't care as much in layout tests?

smfr committed:
[macOS] Make it possible to test smart magnify in layout tests
https://bugs.webkit.org/show_bug.cgi?id=266449
rdar://119699103

Reviewed by Tim Horton.

Make it possible to write a "smart magnify" test. There are several components to this:
1. We teach EventSender how to generate the event, which is an NSEvent of type NSEventTypeSmartMagnify
   with kIOHIDEventTypeZoomToggle.
2. We allow a test to enable WKWebView magnification via UIScriptController. We also fix WebKitTestRunner
   to reset the magnification state between tests.
3. We give WheelEventTestMonitor a defer reason so that a test can wait for the zoom animation to complete
   (yes, this isn't wheel events, but the infrastructure is exactly what we need).
   To support this, DrawingArea::commitTransientZoom() returns an async reply, and
   RemoteLayerTreeDrawingAreaProxyMac::commitTransientZoom() sets up the deferral reason across the
   CAAnimation, and the reply from the web process that the zoom is committed.

A new test is added in LayoutTests/view-gestures/smart-magnify. In future, we should move the swipe
tests into this view-gestures directory.

* LayoutTests/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/resources/ui-helper.js:
(window.UIHelper.async smartMagnifyAt):
(window.UIHelper.setWebViewAllowsMagnification):
* LayoutTests/view-gestures/resources/view-gesture-helpers.js: Added.
(visualViewportDescription):
* LayoutTests/view-gestures/smart-magnify/double-tap-zoom-expected.txt: Added.
* LayoutTests/view-gestures/smart-magnify/double-tap-zoom.html: Added.
* Source/WebCore/page/WheelEventTestMonitor.cpp:
(WebCore::operator<<):
* Source/WebCore/page/WheelEventTestMonitor.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::commitTransientZoom):
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::sendCommitTransientZoom):
* Source/WebKit/UIProcess/mac/TiledCoreAnimationDrawingAreaProxy.mm:
(WebKit::TiledCoreAnimationDrawingAreaProxy::commitTransientZoom):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/DrawingAreaCoordinatedGraphics.cpp:
(WebKit::DrawingAreaCoordinatedGraphics::commitTransientZoom):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/DrawingAreaCoordinatedGraphics.h:
* Source/WebKit/WebProcess/WebPage/DrawingArea.h:
(WebKit::DrawingArea::commitTransientZoom):
* Source/WebKit/WebProcess/WebPage/DrawingArea.messages.in:
* Source/WebKit/WebProcess/WebPage/mac/RemoteLayerTreeDrawingAreaMac.h:
* Source/WebKit/WebProcess/WebPage/mac/RemoteLayerTreeDrawingAreaMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaMac::commitTransientZoom):
* Source/WebKit/WebProcess/WebPage/mac/TiledCoreAnimationDrawingArea.h:
* Source/WebKit/WebProcess/WebPage/mac/TiledCoreAnimationDrawingArea.mm:
(WebKit::TiledCoreAnimationDrawingArea::commitTransientZoom):
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
(WTR::UIScriptController::setWebViewAllowsMagnification):
* Tools/WebKitTestRunner/EventSenderProxy.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/EventSendingController.idl:
* Tools/WebKitTestRunner/InjectedBundle/EventSendingController.cpp:
(WTR::EventSendingController::smartMagnify):
* Tools/WebKitTestRunner/InjectedBundle/EventSendingController.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::didReceiveSynchronousMessageFromInjectedBundle):
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::cocoaResetStateToConsistentValues):
* Tools/WebKitTestRunner/mac/EventSenderProxy.mm:
(-[EventSenderSyntheticEvent initSmartMagnifyEventAtLocation:globalLocation:time:eventNumber:window:]):
(WTR::EventSenderProxy::smartMagnify):
* Tools/WebKitTestRunner/mac/UIScriptControllerMac.h:
* Tools/WebKitTestRunner/mac/UIScriptControllerMac.mm:
(WTR::UIScriptControllerMac::setWebViewAllowsMagnification):

Canonical link: https://commits.webkit.org/272128@main
